File: rust/Card.rs
#!/usr/bin/rust
#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Clone)]
use rand::seq::SliceRandom;
use std::collections::{HashMap, HashSet};
use std::cmp::Ordering;

struct Card {
    suit: char,
    rank: String,
}

// Constructor
impl Card {
    // Constructor
    fn new(rank: &str, suit: char) -> Card {
        Card {
            rank: String::from(rank),
            suit,
        }
    }


 // Method to get the rank of the card
    fn get_rank(&self) -> &str {
        &self.rank
    }
    
// Method to get the suit of the card
    fn get_suit(&self) -> char {
        self.suit
    }



// Method to get the rank value as an integer
    fn get_rank_value(&self) -> i32 {
        match self.rank.as_str() {
            "2" => 2,
            "3" => 3,
            "4" => 4,
            "5" => 5,
            "6" => 6,
            "7" => 7,
            "8" => 8,
            "9" => 9,
            "10" => 10,
            "J" => 11,
            "Q" => 12,
            "K" => 13,
            "A" => 14,
            _ => -1,
        }
    }
}

struct Deck {
    cards: Vec<Cards>,
}

impl Deck {
    fn new() -> Deck {
        let mut cards = Vec::new();
        let suits = ['S', 'H', 'D', 'C'];
        let ranks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"];

        for &suit in suits.iter() {
            for &rank in ranks.iter() {
                cards.push(Cards::new(rank, suit));
            }
        }

        Deck { cards }
    }

    fn shuffle(&mut self) {
        let mut rng = rand::thread_rng();
        self.cards.shuffle(&mut rng);
    }

    fn dealing_cards(&mut self) -> Option<Cards> {
        if !self.cards.is_empty() {
            Some(self.cards.remove(0))
        } else {
            None
        }
    }

    fn size(&self) -> usize {
        self.cards.len()
    }

    fn print_deck(&self) {
        let mut count = 0;

        for card in self.cards.iter() {
            print!("{}\t", card);
            count += 1;
            if count % 13 == 0 {
                println!();
            }
        }
    }

    fn print_remaining_deck(&self) {
        println!("\n*** Here is what remains in the deck...:\t");
        for card in self.cards.iter() {
            print!("{}\t", card);
        }
        println!();
    }
}

#[derive(Debug)]
struct Hand {
    cards: Vec<Cards>,
    hand_rank: String,
}

impl Hand {
    fn new() -> Hand {
        Hand {
            cards: Vec::new(),
            hand_rank: String::from(""),
        }
    }

    fn add_card(&mut self, card: Cards) {
        self.cards.push(card);
    }

    fn get_cards(&self) -> &Vec<Cards> {
        &self.cards
    }

    fn sort_hand(&mut self) {
        self.cards.sort_by(|a, b| a.suit.cmp(&b.suit));
    }

    fn print_hand(&self) {
        let mut count = 0;
        for new_card in self.cards.iter() {
            print!("{}\t", new_card);
            count += 1;
            if count % 5 == 0 {
                println!();
            }
        }
    }

    fn rank_hand(&mut self) {
        let mut suit_counts = HashMap::new();
        let mut rank_counts = HashMap::new();
        self.hand_rank = String::from("Royal Straight Flush");

        for new_card in self.cards.iter() {
            let suit = new_card.suit;
            let rank = &new_card.rank;
            *suit_counts.entry(suit).or_insert(0) += 1;
            *rank_counts.entry(rank.clone()).or_insert(0) += 1;
        }

        if suit_counts.len() == 1 && rank_counts.len() == 5 {
            if self.is_royal_straight_flush(rank_counts.keys().cloned(), suit_counts.keys().next().unwrap().clone()) {
                self.hand_rank = String::from("Royal Straight Flush");
                return;
            }

            self.hand_rank = String::from("Straight Flush");
            return;
        }

        if suit_counts.len() == 1 && rank_counts.len() == 5 {
            if self.is_straight(rank_counts.keys().cloned()) {
                self.hand_rank = String::from("Straight Flush");
                return;
            }

            self.hand_rank = String::from("Flush");
            return;
        }

        if rank_counts.contains_key("4") {
            self.hand_rank = String::from("Four of a Kind");
            return;
        }

        if rank_counts.contains_key("3") && rank_counts.contains_key("2") {
            self.hand_rank = String::from("Full House");
            return;
        }

        if rank_counts.len() == 5 && self.is_straight(rank_counts.keys().cloned()) {
            self.hand_rank = String::from("Straight");
            return;
        }

        if rank_counts.contains_key("3") {
            self.hand_rank = String::from("Three of a Kind");
            return;
        }

        if rank_counts.contains_key("2") {
            let pair_count = rank_counts.values().filter(|&&count| count == 2).count();
            if pair_count == 2 {
                self.hand_rank = String::from("Two Pair");
            } else {
                self.hand_rank = String::from("Pair");
            }
            return;
        }

        self.hand_rank = String::from("High Card");
    }

    fn get_flush_suit(&self) -> char {
        let mut suit_counts = HashMap::new();
        for new_card in self.cards.iter() {
            let new_suit = new_card.suit;
            *suit_counts.entry(new_suit).or_insert(0) += 1;
        }

        for &new_suit in suit_counts.keys() {
            if suit_counts[&new_suit] >= 5 {
                return new_suit;
            }
        }

        ' '
    }

    fn is_royal_straight_flush(&self, ranks: HashSet<String>, suit: char) -> bool {
        ranks.contains("10") && ranks.contains("J") && ranks.contains("Q") && ranks.contains("K") && ranks.contains("A")
    }

    fn is_straight(&self, ranks: HashSet<String>) -> bool {
        let mut rank_values: Vec<i32> = ranks.into_iter().map(|new_rank| Cards::get_rank_value(&new_rank)).collect();
        rank_values.sort();

        let min = rank_values[0];
        let max = *rank_values.last().unwrap();

        (max - min == 4 && rank_values.len() == 5) || (ranks.contains("A") && ranks.contains("2") && ranks.contains("3") && ranks.contains("4") && ranks.contains("5"))
    }
}

#[derive(Debug)]
struct Player {
    name: String,
    hand: Hand,
}

impl Player {
    fn new(name: &str) -> Player {
        Player {
            name: String::from(name),
            hand: Hand::new(),
        }
    }

    fn adding_to_hand(&mut self, new_card: Cards) {
        self.hand.add_card(new_card);
    }

    fn get_name(&self) -> &str {
        &self.name
    }

    fn get_hand(&self) -> &Hand {
        &self.hand
    }
}

impl std::fmt::Display for Player {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

#[derive(Debug)]
struct PlayerRank {
    player: Player,
    rank: String,
    suit: char,
}

impl PlayerRank {
    fn new(player: Player, rank: &str, suit: char) -> PlayerRank {
        PlayerRank {
            player,
            rank: String::from(rank),
            suit,
        }
    }

    fn get_player(&self) -> &Player {
        &self.player
    }

    fn get_rank(&self) -> &str {
        &self.rank
    }

    fn get_suit(&self) -> char {
        self.suit
    }
}

impl Ord for PlayerRank {
    fn cmp(&self, other: &Self) -> Ordering {
        self.rank.cmp(&other.rank)
    }
}

impl PartialOrd for PlayerRank {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Eq for PlayerRank {}

impl PartialEq for PlayerRank {
    fn eq(&self, other: &Self) -> bool {
        self.rank == other.rank
    }
}

impl std::fmt::Display for PlayerRank {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} - {} - Suit: {}", self.player, self.rank, self.suit)
    }
}

fn main() {
    // Create players
    let player1 = Player::new("Player 1");
    let player2 = Player::new("Player 2");

    // Create player ranks
    let player_rank1 = PlayerRank::new(player1.clone(), "Royal Straight Flush", 'S');
    let player_rank2 = PlayerRank::new(player2.clone(), "Straight Flush", 'H');

    // Compare player ranks
    match player_rank1.cmp(&player_rank2) {
        Ordering::Less => println!("Player 1 is ranked lower."),
        Ordering::Greater => println!("Player 1 is ranked higher."),
        Ordering::Equal => println!("Player 1 and Player 2 are ranked equally."),
    }
}

File: julia/Card.jl
using Random

# Cards Type
struct Cards
    rank::String
    suit::String
end

function Cards(rank::String, suit::String)
    return Cards(rank, suit)
end

get_rank(card::Cards) = card.rank
get_suit(card::Cards) = card.suit

Base.:<(card1::Cards, card2::Cards) = get_rank(card1) < get_rank(card2)

Base.show(io::IO, card::Cards) = print(io, get_rank(card), get_suit(card))

function get_rank_value(rank::String)
    rank_values = Dict(
        "2" => 2, "3" => 3, "4" => 4, "5" => 5, "6" => 6, "7" => 7,
        "8" => 8, "9" => 9, "10" => 10, "J" => 11, "Q" => 12, "K" => 13, "A" => 14
    )
    return get(rank_values, rank, -1)
end

# Deck Type
mutable struct Deck
    cards::Vector{Cards}

    function Deck()
        cards = Cards[]
        suits = ['S', 'H', 'D', 'C']
        ranks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]

        for suit in suits
            for rank in ranks
                push!(cards, Cards(rank, suit))
            end
        end

        return new(cards)
    end
end

function shuffle!(deck::Deck)
    Random.shuffle!(deck.cards)
end

function deal_card(deck::Deck)
    if !isempty(deck.cards)
        return popfirst!(deck.cards)
    end
    return nothing
end

size(deck::Deck) = length(deck.cards)

function print_deck(deck::Deck)
    count = 0
    for card in deck.cards
        print(card, "\t")
        count += 1
        if count % 13 == 0
            println()
        end
    end
end

function print_remaining_deck(deck::Deck)
    println("\nRemaining Cards in Deck:\t", "")
    for card in deck.cards
        println(card, "\t")
    end
    println()
end

# Hand Type
mutable struct Hand
    cards::Vector{Cards}
end

function Hand()
    return Hand(Cards[])
end

function add_card(hand::Hand, card::Cards)
    push!(hand.cards, card)
end

function get_cards(hand::Hand)
    return hand.cards
end

function sort_hand!(hand::Hand)
    sort!(hand.cards, by = card -> get_suit(card))
end

Base.:<(hand1::Hand, hand2::Hand) = rank_hand(hand1) < rank_hand(hand2)

function print_hand(hand::Hand)
    count = 0
    for card in hand.cards
        print(card, "\t")
        count += 1
        if count % 5 == 0
            println()
        end
    end
end

Base.show(io::IO, hand::Hand) = join(map(string, hand.cards), "\t")

function rank_hand(hand::Hand)
    suit_counts = Dict()
    rank_counts = Dict()
    hand_rank = "Royal Straight Flush"

    for card in hand.cards
        suit = get_suit(card)
        rank = get_rank(card)
        suit_counts[suit] = get(suit_counts, suit, 0) + 1
        rank_counts[rank] = get(rank_counts, rank, 0) + 1
    end

    if length(suit_counts) == 1 && length(rank_counts) == 5
        if is_royal_straight_flush(keys(rank_counts), first(keys(suit_counts)))
            return "Royal Straight Flush"
        end
        return "Straight Flush"
    end

    if length(suit_counts) == 1 && length(rank_counts) == 5
        if is_straight(keys(rank_counts))
            return "Straight Flush"
        end
        return "Flush"
    end

    if 4 in values(rank_counts)
        return "Four of a Kind"
    end

    if 3 in values(rank_counts) && 2 in values(rank_counts)
        return "Full House"
    end

    if length(rank_counts) == 5 && is_straight(keys(rank_counts))
        return "Straight"
    end

    if 3 in values(rank_counts)
        return "Three of a Kind"
    end

    if 2 in values(rank_counts)
        pair_count = count(x -> x == 2, values(rank_counts))
        if pair_count == 2
            return "Two Pair"
        else
            return "Pair"
        end
    end

    return "High Card"
end

function is_royal_straight_flush(ranks, suit)
    return all(rank in ranks for rank in ["10", "J", "Q", "K", "A"])
end

function is_straight(ranks)
    rank_values = [get_rank_value(rank) for rank in ranks]
    sort!(rank_values)
    min_rank = first(rank_values)
    max_rank = last(rank_values)

    if max_rank - min_rank == 4 && length(ranks) == 5
        return true
    end

    if "A" in ranks && "2" in ranks && "3" in ranks && "4" in ranks && "5" in ranks
        return true
    end

    return false
end

function get_flush_suit(hand::Hand)
    suit_counts = Dict()
    for card in hand.cards
        suit = get_suit(card)
        suit_counts[suit] = get(suit_counts, suit, 0) + 1
    end

    for (suit, count) in suit_counts
        if count >= 5
            return suit
        end
    end

    return nothing  # No flush suit found
end

# Player Type
mutable struct Player
    name::String
    hand::Hand
end

function Player(name::String)
    return Player(name, Hand())
end

function add_to_hand(player::Player, card::Cards)
    add_card(player.hand, card)
end

get_name(player::Player) = player.name

get_hand(player::Player) = player.hand

Base.show(io::IO, player::Player) = player.name

# PlayerRank Type
mutable struct PlayerRank
    player::Player
    rank::String
    suit::String
end

get_player(player_rank::PlayerRank) = player_rank.player

get_rank(player_rank::PlayerRank) = player_rank.rank

Base.:<(player_rank1::PlayerRank, player_rank2::PlayerRank) = get_rank(player_rank1) < get_rank(player_rank2)

get_suit(player_rank::PlayerRank) = player_rank.suit